### PR TITLE
hir: Disallow `target_feature` on constants

### LINKED ISSUE #64768: ICE when using target_feature with a const
The following code snippet triggers a ICE:
```rust
#[target_feature(avx2)]
const FOO: usize = 7;
```

`rustc 1.37.0 (eae3437df 2019-08-13) running on x86_64-unknown-linux-gnu`

Bug was found on playground

### PR DESCRIPTION
**Pull request by davidtwco**
Fixes #64768.

This PR fixes an ICE when `#[target_feature]` is applied to constants by disallowing this with the same error as when `#[target_feature]` is applied to other places it shouldn't be. 

I couldn't see anything in the [RFC](https://github.com/rust-lang/rfcs/blob/master/text/2045-target-feature.md) that suggested that `#[target_feature]` should be applicable to constants or any tests that suggested it should, though I might have missed something - if this is desirable in future, it remains possible to remove this error (but for the time being, I think this error is better than an ICE).

I also added some extra cases to the test for other places where `#[target_feature]` should not be permitted.

cc @gnzlbg

### CODE REVIEW

File: src/librustc/hir/check_attr.rs [hunk 1]
```diff
@@ -93,9 +93,10 @@ struct CheckAttrVisitor<'tcx> {
 impl CheckAttrVisitor<'tcx> {
     /// Checks any attribute.
     fn check_attributes(&self, item: &hir::Item, target: Target) {
-        if target == Target::Fn || target == Target::Const {
+        let target_feature = item.attrs.iter().find(|a| a.check_name(sym::target_feature));
+        if target == Target::Fn || (target == Target::Const && target_feature.is_none()) {
             self.tcx.codegen_fn_attrs(self.tcx.hir().local_def_id(item.hir_id));
```
Comment (varkor) [inline src/librustc/hir/check_attr.rs]: Why do we call `codegen_fn_attrs` at all for consts? Most of the codegen attributes don't make any sense for constants. If any are allowed on consts, we should probably have separate handling for them. Only `ffi_returns_twice` seems to throw an error for non-function items.

[This PR added the `target == Target::Const`](https://github.com/rust-lang/rust/pull/51966/files#diff-a7a4db27e720b5572001d15719153b54), but without any comment as to why.

Comment (varkor) [reply src/librustc/hir/check_attr.rs]: I think it makes sense to move the validity checks below above `codegen_fn_attrs` and early-return if any of them fail. Then the check for `target_feature` can fit into the same loop. We should also have checks for the other, missing, attributes, but that could happen in another PR.

**Comment by gnzlbg** [comment]
This LGTM, target feature should only be allowed on functions and functions
only - I think we should disallow it for const fns.

File: src/librustc/hir/check_attr.rs (hunk 1 above)
Comment (davidtwco) [reply src/librustc/hir/check_attr.rs]: I've changed the PR to what I think you've suggested here.

File: src/librustc/hir/check_attr.rs [hunk 2]
```diff
@@ -93,70 +93,91 @@ struct CheckAttrVisitor<'tcx> {
...
                 self.check_non_exhaustive(attr, item, target)
             } else if attr.check_name(sym::marker) {
                 self.check_marker(attr, item, target)
+            } else if attr.check_name(sym::target_feature) {
+                self.check_target_feature(attr, item, target)
+            } else {
+                true
+            };
+
+            if !is_valid {
```
Comment (varkor) [inline src/librustc/hir/check_attr.rs]: Could you move this check outside the loop, and add a test case for reporting errors for multiple invalid attributes (if there's not one already)?

**Review by varkor — CHANGES_REQUESTED** [review]
Thanks! r=me after a test case for multiple invalid attributes.

File: src/librustc/hir/check_attr.rs (hunk 2 above)
Comment (davidtwco) [reply src/librustc/hir/check_attr.rs]: I've moved the check, but I haven't added a test because when I did (with two invalid attributes), only one error was reported - I don't think it's my changes doing that.

Comment (varkor) [reply src/librustc/hir/check_attr.rs]:
```rust
#![feature(marker_trait_attr)]
#![feature(non_exhaustive)]

#[marker]
#[non_exhaustive]
const FOO: u8 = 0;
```
fails with two errors on [the playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=27edd1b3f033041c69225396b7c15a11). Does it not work if you stop short-circuiting with `is_valid = is_valid && [...]`?

File: src/librustc/hir/check_attr.rs [hunk 3]
```diff
@@ -93,70 +93,102 @@ struct CheckAttrVisitor<'tcx> {
...
+            } else {
+                true
+            };
+        }
+
+        if !is_valid {
+            return;
+        }
+
+        if target == Target::Fn || target == Target::Const {
```
Comment (varkor) [inline src/librustc/hir/check_attr.rs]: I think we might actually be able to [get rid of the check](https://github.com/rust-lang/rust/pull/51966/files/42eb85002ae4bda9899216805d03fa7d77279ede#r328653468) for `Target::Const` completely, now that `wasm_custom_section` is no longer a thing. But we can try that in a follow up PR.

**Comment by moonheart08** [comment]
Of note: The reason I was attempting to use target function on a const was because the const needed a different value on different targets.
For AVX2 targets, it needed a value of 7, for AVX512 targets, a value of 15, and for other targets, a value of 3. Pointers on how to do it right are appreciated.

File: src/librustc/hir/check_attr.rs [hunk 4]
```diff
@@ -93,70 +93,102 @@ struct CheckAttrVisitor<'tcx> {
...
-        if target == Target::Fn || target == Target::Const {
-            self.tcx.codegen_fn_attrs(self.tcx.hir().local_def_id(item.hir_id));
-        } else if let Some(a) = item.attrs.iter().find(|a| a.check_name(sym::target_feature)) {
-            self.tcx.sess.struct_span_err(a.span, "attribute should be applied to a function")
-                .span_label(item.span, "not a function")
-                .emit();
-        }
-
+        let mut is_valid = true;
         for attr in &item.attrs {
```
Comment (Centril) [inline src/librustc/hir/check_attr.rs]: While we're at it... Maybe reformulate this with `item.attrs.all(...)`?

File: src/test/ui/target-feature-wrong.rs [hunk 5]
```diff
@@ -35,6 +35,31 @@ fn bar() {}
 mod another {}
 //~^ NOTE not a function
 
+#[target_feature(enable = "sse2")]
```
Comment (Centril) [inline src/test/ui/target-feature-wrong.rs]: Drive-by: can we move the test to not be right in `src/test/ui/` ; that's ungreat organization :)

File: src/librustc/hir/check_attr.rs [hunk 6]
```diff
@@ -93,70 +93,104 @@ struct CheckAttrVisitor<'tcx> {
...
-            if attr.check_name(sym::inline) {
+            let current_attr_is_valid = if attr.check_name(sym::inline) {
```
Comment (varkor) [inline src/librustc/hir/check_attr.rs]:
```suggestion
            is_valid &= if attr.check_name(sym::inline) {
```

File: src/librustc/hir/check_attr.rs (hunk 4 above)
Comment (varkor) [reply src/librustc/hir/check_attr.rs]: I personally prefer the explicit loop here.

File: src/librustc/hir/check_attr.rs (hunk 6 above)
Comment (Centril) [reply src/librustc/hir/check_attr.rs]: Imo the current formulation is cleaner because it is amenable to being expressed as a fold (and avoiding `let mut`).

**Comment by varkor** [comment]
r=me when tests pass (with or without the final suggestions, which are entirely preferential).

File: src/librustc/hir/check_attr.rs (hunk 6 above)
Comment (Centril) [reply src/librustc/hir/check_attr.rs]: (but `.all` is short-circuiting, which we might not want...)

Comment (varkor) [reply src/librustc/hir/check_attr.rs]: We don't: we want to emit all errors.

File: src/test/ui/target-feature-wrong.rs (hunk 5 above)
Comment (davidtwco) [reply src/test/ui/target-feature-wrong.rs]: Fixed, thanks.

File: src/librustc/hir/check_attr.rs (hunk 4 above)
Comment (davidtwco) [reply src/librustc/hir/check_attr.rs]: I've not changed this because `.all(...)` stops as soon as the first attribute returns false, so we don't report multiple errors if I make this change.

File: src/librustc/hir/check_attr.rs (hunk 3 above)
Comment (davidtwco) [reply src/librustc/hir/check_attr.rs]: I've removed this in a separate commit.

File: src/librustc/hir/check_attr.rs (hunk 2 above)
Comment (davidtwco) [reply src/librustc/hir/check_attr.rs]: Oops, you're right, forgot that it would short-circuit.

**Comment by gnzlbg** [comment]
@moonheart08

> Of note: The reason I was attempting to use target function on a const
was because the const needed a different value on different targets.
For AVX2 targets, it needed a value of 7, for AVX512 targets, a value of
15, and for other targets, a value of 3. Pointers on how to do it right are
appreciated.

You can use #[cfg(target_feature)] for that.

**Comment by moonheart08** [comment]
> @moonheart08
> Of note: The reason I was attempting to use target function on a const
> was because the const needed a different value on different targets. For AVX2 targets, it needed a value of 7, for AVX512 targets, a value of 15, and for other targets, a value of 3. Pointers on how to do it right are appreciated. You can use #[cfg(target_feature)] for that.

I had deleted my comment and moved it to the issue that caused this PR, thank you, however!
